README.md
# metrics-raw-kairosdb
Implements the EventListener Interface defined in metrics-raw.  Sends metric-event data to KairosDB.

# Usage

Creating a KairosDBListener via KairosDBListenerFactory:
1.  buildListener(String connectString, MetricRegistry registry)
2. public static EventListener buildListener(String connectString,
                                             String username,
                                             String password,
                                             MetricRegistry registry,
                                             int batchSize,
                                             int bufferSize,
                                             long offerTimeMillis,
                                             int maxDispatchThreads)

Where:
 1. batchSize = max events sent to the kairosdb per http request.
 2. bufferSize = max internal buffer
 3. maxDispatchThreads = max number of concurrent httpRequests/threads for sending events to kairosdb.





ReleaseNotes.txt
0.6.0
 * Improved configuration, allows multi-threaded event 'flush' to kairosdb. 
 * Breaking Build - KairosDBListenerFactory.buildListener updated.
 * Updated kairosDBListenerFactory - buildListener now allows batchSize,bufferSize,maxDispatchThreads to be configured.
0.5.3
 * When the kairosdblistener emits statistics, it will not report namespace,app,app-type, as this became problematic when multiple namespace,app,app-type are reported with the same Kairosdb-listener.
0.5.0
 * update to use latest metrics-raw.
0.4.6
* Version bump for version-parity with metrics-raw 
0.4.4
* updated to reference latest metrics-raw.
* updated to enable use of the KairosDB*LIsteners without the metricRegistry. (i.e. registry is nullable).
* update references of serviceTeam to namespace
0.0.2
* added metrics version info to metrics status, under metricsraw.stats.data.count (2 new tags: kairDbListenerVersion, and metricsRawVersion)



src/main/java/org/dshops/test/generators/UtilArg.java
package org.dshops.test.generators;

public class UtilArg {
    static public int getIntArg(String[] args, String arg, int defaultValue) {
        String x = getArg(args, arg, defaultValue + "");
        return Integer.parseInt(x);
    }

    static public String getArg(String[] args, String arg, String defaultValue) {
        String match = "-" + arg + "=";
        for (String x : args) {
            if (x != null && x.startsWith(match)) {
                return x.substring(match.length());
            }
        }
        return defaultValue;
    }

}


src/main/java/org/dshops/test/metrics/generators/UtilArg.java
package org.dshops.test.metrics.generators;

public class UtilArg {
    static public int getIntArg(String[] args, String arg, int defaultValue) {
        String x = getArg(args, arg, defaultValue + "");
        return Integer.parseInt(x);
    }

    static public String getArg(String[] args, String arg, String defaultValue) {
        String match = "-" + arg + "=";
        for (String x : args) {
            if (x != null && x.startsWith(match)) {
                return x.substring(match.length());
            }
        }
        return defaultValue;
    }

}


src/test/java/org/dshops/test/generators/MetricGenerator.java
package org.dshops.test.generators;

import static org.dshops.test.generators.UtilArg.getArg;
import static org.dshops.test.generators.UtilArg.getIntArg;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import org.dshops.listeners.KairosDBListener;
import org.dshops.metrics.EventListener;
import org.dshops.metrics.MetricRegistry;
import org.dshops.test.generators.support.EventGenerator;
import org.dshops.test.generators.support.EventQueryGenerator;
import org.kairosdb.client.HttpClient;

/** Metric Generators is used for 'testing', aka generating metrics
 *  This generator is specific to KairosDB., and will be moved library metric-raw-kairosdb in the future.
 * */
public class MetricGenerator {

    public static void main(String[] args) {
        new MetricGenerator(args);
    }

    public EventListener getListener(MetricRegistry reg, String url) {
        return new KairosDBListener(url,
                                    "root",
                                    "root",
                                    reg,
                                    100);
    }

    public MetricGenerator(String[] args) {
        try {
            String url = getArg(args, "url", "http://localhost:8080");
            String user = getArg(args, "u", "root");
            String pd = getArg(args, "p", "root");

            String service = getArg(args, "s", "dshops");
            String app = getArg(args, "a", "metrics");
            String appType = getArg(args, "T", "test");

            // simulations
            final long runTime = getIntArg(args, "t", Integer.MAX_VALUE); // runtime minutes
            int maxEvents = getIntArg(args,"m", Integer.MAX_VALUE);       // maximum number of events to generate )-1 unlimited, test will exist at the time limit.
            final int hosts = getIntArg(args, "h", 1);                    // unique host tag names: host<#>   (1 thread per host ( 1 MR per host)

            // values below are 'per-thread'
            final int eventSignatures = getIntArg(args, "e", 1);    // unique Event Names/thread        (event<#>)
            final int tagCount = getIntArg(args, "tagCount", 1);    // even event generated will have a random number of tags (upto this number)
            final int tagValues = getIntArg(args, "tagValues", 5);  // random value per tag
            final int writeTps = getIntArg(args, "tps", 500);       // tps target per host
            final int querytps = getIntArg(args, "qtps", 0);        // tps target per host

            EventGenerator[] writers = new EventGenerator[hosts];
            AtomicInteger dpWrittenCount = new AtomicInteger();
            AtomicInteger queryCounter = new AtomicInteger();
            AtomicInteger dataPointsRead = new AtomicInteger();
            AtomicBoolean exitFlag = new AtomicBoolean();
            System.out.println("SETTINGS");
            System.out.println("url                 " + url);
            System.out.println("service             " + service);
            System.out.println("app                 " + app);
            System.out.println("app                 " + appType);
            System.out.println("runtime             " + runTime);
            System.out.println("eventNames/thread:  " + eventSignatures);
            System.out.println("tags/thread:        " + tagCount);
            System.out.println("values/thread:      " + tagValues);
            System.out.println("TPS/thread:         " + writeTps);
            System.out.println("Query-TPS/thread:   " + querytps);


            for (int i = 0; i < hosts; i++) {
                String hostname = "host"+i;
                MetricRegistry mr = new MetricRegistry.Builder(service, app, appType, "hostname", "datacenter1").build();

                if (writeTps > 0) {
                    writers[i] = new EventGenerator(hostname,
                                                eventSignatures,
                                                exitFlag,
                                                dpWrittenCount,
                                                mr,
                                                tagCount,
                                                tagValues,
                                                writeTps);
                }

                String[] eventNames = new String[eventSignatures];
                for (int j = 0 ; j < eventNames.length;j++){
                    eventNames[j] = "event"+j;
                }

                String[] tagVals = new String[tagValues];
                for (int k = 0; k < tagVals.length; k++){
                    tagVals[k] = "value"+k;
                }

                Map<String,String[]> tags = new HashMap<>();
                for (int j = 0; j < tagCount; j++) {
                    tags.put("tag"+j, tagVals);
                }
                HttpClient kairosDb = new HttpClient(url);
                EventQueryGenerator eqg = new EventQueryGenerator(kairosDb,
                                                                  eventNames,
                                                                  tags,
                                                                  querytps,
                                                                  exitFlag,
                                                                  service+"."+app, // prefix
                                                                  queryCounter,
                                                                  dataPointsRead);
                eqg.start();

                if (writeTps > 0) {
                    getListener(mr, url);
                    writers[i].start();
                }
            }
            long endTime = System.currentTimeMillis() + (runTime * 60_000);
            long reportTime = System.currentTimeMillis() + 1_000;
            long generatedEvents = 0;
            long queries = 0;
            while (System.currentTimeMillis() < endTime && dpWrittenCount.get() < maxEvents) {
                Thread.sleep(1000); // 10 seconds
                if (System.currentTimeMillis() > reportTime) {
                    long remainingTime = endTime - System.currentTimeMillis();

                    long newEvents = dpWrittenCount.get() - generatedEvents;
                    float writtenTps = newEvents / 10;
                    generatedEvents += newEvents;

                    long newQueries = queryCounter.get() - queries;
                    float qtps = newQueries / 10;
                    queries += newQueries;
                    System.out.println("-->WRITES--[dataPoints:" + dpWrittenCount.get() + " TPS:" +writtenTps+ "] READS--[queries:"+queryCounter+" QPS:"+qtps+" dataPoints:"+dataPointsRead+"]  Time Remaining (sec): " + (remainingTime/1000));
                    reportTime = System.currentTimeMillis() + 10_000;
                }
            }
            exitFlag.set(true);
        }
        catch(Exception e) {
            e.printStackTrace();
        }

    }


}



src/main/java/org/dshops/metrics/listeners/KairosDbHybridListener.java
package org.dshops.metrics.listeners;

import java.util.HashSet;
import java.util.Set;

import org.dshops.metrics.Event;
import org.dshops.metrics.MetricRegistry;

public class KairosDbHybridListener extends ThreadedListener {

    private final KairosDbListenerMilliBucket bucketListener;
    private final KairosDBListener indexedListener;
    private Set<String> indexedMetrics = new HashSet<>();
    private final MetricRegistry registry;

    KairosDbHybridListener(String connectString, MetricRegistry registry) {
        this.registry = registry;
        bucketListener = new KairosDbListenerMilliBucket(connectString, "", "", registry);
        indexedListener = new KairosDBListener(bucketListener.getClient(), registry);
    }

    public void setIndexedMetrics(Set<String> metricNames) {
        indexedMetrics.clear();
        metricNames.stream()
                    .forEach( metricName ->
                        indexedMetrics.add(registry.getPrefix() + metricName));
    }

    @Override
    public void onEvent(Event e) {
        if (indexedMetrics.contains(e.getName())){
            indexedListener.onEvent(e);
        }
        else {
            bucketListener.onEvent(e);
        }
    }

    @Override
    public int eventsBuffered() {
        return bucketListener.eventsBuffered() + indexedListener.eventsBuffered();
    }

    @Override
    public void stop() {
        bucketListener.stop();
        indexedListener.stop();
    }

}


src/main/java/org/dshops/metrics/listeners/KairosDBListenerFactory.java
package org.dshops.metrics.listeners;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.dshops.metrics.EventListener;
import org.dshops.metrics.MetricRegistry;

public class KairosDBListenerFactory {
    private static Map<String, EventListener> indexingListeners = new ConcurrentHashMap<>();
    private static Map<String, EventListener> unIndexingListeners = new ConcurrentHashMap<>();
    private static boolean enableListenerCaching = true;

    public static void enableListenerCaching(boolean enableCaching) {
        enableListenerCaching = enableCaching;
    }

    public static boolean isListenerCachingEnabled() {
        return enableListenerCaching;
    }

    public static EventListener buildListener(String connectString, MetricRegistry registry) {
        return buildListener(connectString, "", "", registry, 50, 5000, -1, 2);
    }

    public static EventListener buildListener(String connectString,
                                               String username,
                                               String password,
                                               MetricRegistry registry,
                                               int batchSize,
                                               int bufferSize,
                                               long offerTimeMillis,
                                               int maxDispatchThreads) {
        EventListener listener = indexingListeners.get(connectString);
        if (listener == null) {
            synchronized (indexingListeners) {
                listener = indexingListeners.get(connectString);
                if (listener == null) {
                    listener = new KairosDbIndexingListener(connectString,
                                                            username,
                                                            password,
                                                            registry,
                                                            batchSize,
                                                            bufferSize,
                                                            offerTimeMillis,
                                                            maxDispatchThreads);
                    if (enableListenerCaching) {
                        indexingListeners.put(connectString, listener);
                    }
                }
            }
        }
        return listener;
    }

    public static EventListener buildUnindexedListener(String connectString, MetricRegistry registry) {
        return buildUnindexedListener(connectString, "", "", registry, 50, 5000, -1);
    }

    public static EventListener buildUnindexedListener(String connectString,
                                                       String username,
                                                       String password,
                                                       MetricRegistry registry,
                                                       int batchSize,
                                                       int bufferSize,
                                                       long offerTimeMillis) {
        EventListener listener = unIndexingListeners.get(connectString);
        if (listener == null) {
            synchronized (unIndexingListeners) {
                listener = unIndexingListeners.get(connectString);
                if (listener == null) {
                    listener = new KairosDbNonIndexingListener(connectString,
                                                               username,
                                                               password,
                                                               registry,
                                                               batchSize,
                                                               bufferSize,
                                                               offerTimeMillis);
                    if(enableListenerCaching) {
                        unIndexingListeners.put(connectString, listener);
                    }
                }
            }
        }
        return listener;
    }
}


src/main/java/org/dshops/metrics/listeners/KairosDBCountingListener.java
package org.dshops.metrics.listeners;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.LongAdder;

import org.dshops.metrics.Event;
import org.dshops.metrics.EventListener;
import org.dshops.metrics.MetricKey;

/**
 * Notes: while all constructors take username/password, these are currently not used.
 *
 * Upload strategy:
 * All events will be uploaded in batch: batchSize (default: 100), or every 1 second, whichever comes first.
 *
 * */
public class KairosDBCountingListener extends ThreadedListener implements Runnable {
    private EventListener listener;
    private Map<MetricKey,ResetCounter> counts = new ConcurrentHashMap<>();

    KairosDBCountingListener(EventListener listener) {
        this.listener = listener;
    }

    @Override
    public void onEvent(Event e) {
        ResetCounter counter = counts.get(e.getHash());
        if (counter == null) {
            synchronized (counts) {
                counter = counts.get(e.getHash());
                if (counter == null) {
                    counter = new ResetCounter();
                    counts.put(e.getHash(),counter);
                }
            }
        }
        try {
            e.setIndex(counter.getCount());
        }catch(Exception ex){
            ex.printStackTrace();
        }
        listener.onEvent(e);
    }

    @Override
    public int eventsBuffered() {
        // TODO Auto-generated method stub
        return 0;
    }

    @Override
    public void run() {
        // TODO Auto-generated method stub

    }
}

// option 1 (no gc overhead/atomic/sync)
class ResetCounter {
    public AtomicInteger counter = new AtomicInteger();
    public long ts;
    public int getCount() {
        int count = counter.incrementAndGet();
        if (System.currentTimeMillis() - ts  > 1) {
            synchronized (this) {
                ts = System.currentTimeMillis();
                counter.set(0);
            }
        }
        return count;
    }
}

//option 2 (gc overhead), no atomics
class ResetCounter2 {
 public LongAdder counter;
 public long ts;

 public int getCount() {
     counter.increment();
     int count = counter.intValue();
     if (System.currentTimeMillis() - ts  > 1) {
         ts = System.currentTimeMillis();
         counter = new LongAdder();
     }
     return count;
 }
}

src/main/java/org/dshops/metrics/listeners/KairosDBListener.java
package org.dshops.metrics.listeners;

import java.io.InputStream;
import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.concurrent.atomic.AtomicInteger;

import org.dshops.metrics.DoubleEvent;
import org.dshops.metrics.Event;
import org.dshops.metrics.LongEvent;
import org.dshops.metrics.MetricKey;
import org.dshops.metrics.MetricRegistry;
import org.kairosdb.client.HttpClient;
import org.kairosdb.client.builder.MetricBuilder;
import org.kairosdb.client.response.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Notes: while all constructors take username/password, these are currently not used.
 *
 * Upload strategy:
 * All events will be uploaded in batch: batchSize (default: 100), or every 1 second, whichever comes first.
 *
 * */
public class KairosDBListener extends ThreadedListener implements Runnable {
    private final ConcurrentSkipListMap<Long,MetricTimeSlot> metricBuffer = new ConcurrentSkipListMap<>(); // time/
    private final int batchSize;
    private final long offerTime;   // amount of time we are willing to 'block' before adding an event to our buffer, prior to dropping it.
    private final HttpClient kairosDb;
    private final static Logger log = LoggerFactory.getLogger(KairosDBListener.class);
    private final MetricRegistry registry;
    private final AtomicInteger droppedEvents = new AtomicInteger();
    private final String serviceTeam;
    private final String app;
    private final String appType;
    private final int bufferLimit;
    private final AtomicInteger bufferedEvents = new AtomicInteger();
    private final Map<String,String> versions = new HashMap<>();

    KairosDBListener(String connectString,
                            String un,
                            String pd,
                            MetricRegistry registry) {
        this(connectString, un, pd, registry, 100);
    }

    KairosDBListener(String connectString,
                            String un,
                            String pd,
                            MetricRegistry registry,
                            int batchSize) {
        this(connectString, un, pd, registry, batchSize, 5000, -1);
    }
    KairosDBListener(HttpClient client,
            MetricRegistry registry) {
        this(client, registry, 100, 5000, -1);
    }
    KairosDBListener(HttpClient client,
                     MetricRegistry registry,
                     int batchSize,
                     int bufferSize,
                     long offerTimeMillis) {
        this.registry = registry;
        String[] prefix = registry.getPrefix().split("\\.");
        this.serviceTeam = prefix[0];
        this.app = prefix[1];
        this.appType = prefix[2];
        this.bufferLimit = bufferSize;

        if (batchSize > 1) {
            this.batchSize = batchSize;
        }
        else {
            this.batchSize = 100;
        }

        this.offerTime = offerTimeMillis;
        this.kairosDb = client;
        startThread(registry);
    }

    KairosDBListener(String connectString,
                            String un,
                            String pd,
                            MetricRegistry registry,
                            int batchSize,
                            int bufferSize,
                            long offerTimeMillis) {
        this.registry = registry;
        String[] prefix = registry.getPrefix().split("\\.");
        this.serviceTeam = prefix[0];
        this.app = prefix[1];
        this.appType = prefix[2];
        this.bufferLimit = bufferSize;

        if (batchSize > 1) {
            this.batchSize = batchSize;
        }
        else {
            this.batchSize = 100;
        }

        this.offerTime = offerTimeMillis;
        try {
            if (connectString != null) {
                // todo: no way configure timeout?
                kairosDb = new HttpClient(connectString);
            }
            else {
                log.warn("kairosDb Url is null - running in noop mode.");
                kairosDb = null;
            }
        }
        catch(MalformedURLException mue) {
            throw new RuntimeException("Malformed Url:"+connectString+" "+mue.getMessage());
        }
        startThread(registry);
    }

    private void startThread(MetricRegistry registry) {
        String kbListenerVersion = getVersion("org.dshops/metrics-raw-kairosdb", this.getClass());
        String metricsRawVersion = getVersion("org.dshops/metrics-raw", registry.getClass());

        System.out.println("kairosDbListener - Version Info[KairosDbListener:" + kbListenerVersion + ", metrics-raw:"+metricsRawVersion);

        if (kbListenerVersion != null) {
            versions.put("kairosDbListenerVersion", kbListenerVersion);
            versions.put("metricsRawVersion", metricsRawVersion);
        }

        runThread = new Thread(this);
        runThread.setName("kairosDbListener");
        runThread.setDaemon(true);
        runThread.start();
    }

    @Override
    public void run() {
        final List<IndexedEvent> dispatchList = new ArrayList<>(batchSize);
        long lastResetTime = System.currentTimeMillis();
        long httpCalls = 0;
        long metricCount = 0;
        long errorCount = 0;
        long exceptionTime = 0;
        Response lastError = null;
        do {
            try {
                addNewEvents(dispatchList);
                if(!dispatchList.isEmpty()) {
                    // @todo - consider bucketing..we may get multiple datapoints for the same ms, with the same name/tagSet, we will lose data with this approach atm.
                    metricCount += dispatchList.size();
                    if (kairosDb != null) {
                        Response r = kairosDb.pushMetrics(buildPayload(dispatchList));
                        httpCalls++;
                        if (r.getStatusCode() != 204 ) {
                            lastError = r;
                            errorCount++;
                        }
                    }
                    // Generate stat info
                    if (System.currentTimeMillis() - lastResetTime > 60_000) {
                        sendMetricStats(metricCount, errorCount, httpCalls);

                        if (lastError != null) {
                            StringBuilder sb = new StringBuilder();
                            for (String s : lastError.getErrors()) {
                                sb.append("[");
                                sb.append(s);
                                sb.append("]");
                            }
                            if (lastError != null) {
                                log.error("Http calls:{} Dispatch count: {} errorCount:{} lastError.status:{} lastErrorDetails:{}", httpCalls, metricCount, errorCount, lastError.getStatusCode(), sb.toString());
                            }
                            lastError = null;
                        }
                        droppedEvents.set(0);
                        httpCalls = 0;
                        errorCount = 0;
                        metricCount = 0;
                        lastResetTime = System.currentTimeMillis();
                    }
                }
            }
            catch(InterruptedException ie) {
                break;
            }
            catch(Exception ex) {
                errorCount++;
                if (System.currentTimeMillis() - exceptionTime > 60_000) {
                    log.error("Unexpected Exception (only 1 exception logged per minute)", ex);
                }
                exceptionTime = System.currentTimeMillis();
            }
            finally {
                dispatchList.clear();
            }
        } while(!stopRequested);
    }

    @Override
    public void stop() {
        super.stop();
        if(kairosDb != null) {
            try {
                kairosDb.shutdown();
            }
            catch (Exception e) {}
        }
    }

    // move events from the MetricBuffer to the dispatchList, will exist if over 1000 second of processing
    // or we have exceeded batchSize.
    // Any timeSlot started, will result in all events from that timeslot being added to dispatchList.
    private void addNewEvents(final List<IndexedEvent> dispatchList) throws InterruptedException {
        long startTime = System.currentTimeMillis();
        do {
            Long ts = null;
            Map.Entry<Long,MetricTimeSlot> entry = metricBuffer.firstEntry();
            if (entry != null) {
                ts = entry.getKey();
                if (ts != null && ts < System.currentTimeMillis() - 50) {
                    MetricTimeSlot timeSlot = metricBuffer.remove(ts);
                    for (Map.Entry<MetricKey, ConcurrentLinkedQueue<Event>> metricEvents : timeSlot.metricMap.entrySet()) {
                        int i = 0;
                        for (Event event: metricEvents.getValue()) {
                            dispatchList.add(new IndexedEvent(event, i++));
                            bufferedEvents.decrementAndGet();
                        }
                    }
                    // send what we have if we have exceeded batchSize, or have gone over 1 second.
                    if (dispatchList.size() > batchSize || System.currentTimeMillis() - startTime > 1000) {
                        break;
                    }
                }
            }
            else {
                Thread.sleep(50);
            }
        } while(!stopRequested);
    }

    private void sendMetricStats(long metricCount, long errorCount, long httpCalls) throws Exception {
        if (kairosDb == null) return;
        try {
            MetricBuilder mb = MetricBuilder.getInstance();


            mb.addMetric("metricsraw.stats.data.count")
              .addTags(registry.getTags())
              .addTags(versions)
              .addTag("serviceTeam",serviceTeam)
              .addTag("app",app)
              .addTag("appType",appType)
              .addDataPoint(metricCount);
            mb.addMetric("metricsraw.stats.http.errors")
            .addTags(registry.getTags())
            .addTag("serviceTeam",serviceTeam)
            .addTag("app",app)
            .addTag("appType",appType)
              .addDataPoint(errorCount);
            mb.addMetric("metricsraw.stats.http.count")
            .addTags(registry.getTags())
            .addTag("serviceTeam",serviceTeam)
            .addTag("app",app)
            .addTag("appType",appType)
              .addDataPoint(httpCalls);
            mb.addMetric("metricsraw.stats.data.dropped")
            .addTags(registry.getTags())
            .addTag("serviceTeam",serviceTeam)
            .addTag("app",app)
            .addTag("appType",appType)
              .addDataPoint(droppedEvents.longValue());

             Response r = kairosDb.pushMetrics(mb);
             if (r.getStatusCode() != 204 ) {
                 log.warn("failed to send metric statistics!", r.getStatusCode());
             }
        }
        catch(Exception e) {
            log.warn("failed to send metric statistis to server! {} ", e.getMessage());
        }
    }

    private MetricBuilder buildPayload(List<IndexedEvent> events) {
        MetricBuilder mb = MetricBuilder.getInstance();
        for (IndexedEvent ie: events) {
            Event e = ie.event;
            if (e instanceof LongEvent) {
                mb.addMetric(e.getName())
                              .addTags(e.getTags())
                              .addTag("index", ie.index+"")
                              .addDataPoint(e.getTimestamp(), e.getLongValue());
            }
            else if (e instanceof DoubleEvent) {
                mb.addMetric(e.getName())
                              .addTags(e.getTags())
                              .addTag("index", ie.index+"")
                              .addDataPoint(e.getTimestamp(), e.getDoubleValue());
            }
            else {
                // this is a pure event, value has no meaning
                mb.addMetric(e.getName())
                              .addTags(e.getTags())
                              .addTag("index", ie.index+"")
                              .addDataPoint(e.getTimestamp(), 1);
            }
        }
        return mb;
    }

    @Override
    public int eventsBuffered() {
        return bufferedEvents.get();
    }

    @Override
    public void onEvent(Event e) {

        if (bufferedEvents.get() > bufferLimit) {
            if (offerTime > 0) {
                try {
                    long startTime = System.currentTimeMillis();
                    while (bufferedEvents.get() > bufferLimit && System.currentTimeMillis() - startTime < offerTime) {
                        Thread.sleep(10);
                    }
                    if (bufferedEvents.get() > bufferLimit){
                        droppedEvents.incrementAndGet();
                        return;
                    }
                }
                catch(InterruptedException ie) {
                    return;
                }
            }
            else {
                droppedEvents.incrementAndGet();
                return;
            }
        }

        MetricTimeSlot timeSlot = metricBuffer.get(e.getTimestamp());
        if (timeSlot == null) {
            timeSlot = new MetricTimeSlot();
            metricBuffer.put(e.getTimestamp(), timeSlot);
        }
        ConcurrentLinkedQueue<Event> events = timeSlot.metricMap.get(e.getHash());
        if (events == null) {
            events = new ConcurrentLinkedQueue<>();
            timeSlot.metricMap.put(e.getHash(), events);
            bufferedEvents.incrementAndGet();
        }
        events.add(e);
    }

    public synchronized String getVersion(String path, Class clazz) {
        String version = null;

        // try to load from maven properties first
        try {
            Properties p = new Properties();
            InputStream is = getClass().getResourceAsStream("/META-INF/maven/"+path+"/pom.properties");
            if (is != null) {
                p.load(is);
                version = p.getProperty("version", "");
            }
        }
        catch (Exception e) {
            // ignore
        }

        // fallback to using Java API
        if (version == null) {
            Package aPackage = clazz.getPackage();
            if (aPackage != null) {
                version = aPackage.getImplementationVersion();
                if (version == null) {
                    version = aPackage.getSpecificationVersion();
                }
            }
        }

        if (version == null) {
            // we could not compute the version so use a blank
            version = "";
        }

        return version;
    }
}

class IndexedEvent {
    final Event event;
    final int index;
    public IndexedEvent(Event e, int index){
        this.event = e;
        this.index = index;
    }
}

class MetricTimeSlot {
    Map<MetricKey,ConcurrentLinkedQueue<Event>> metricMap = new ConcurrentHashMap<>();
}
